core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java

```diff
@@ -0,0 +1,275 @@
+        for (int i = 0; i < builders.length; i++) {
+            builders[i] = client().prepareIndex(index, "type").setSource("field", "value");
+        }
+        indexRandom(true, builders);
+        client().admin().indices().prepareFlush().setForce(true).setWaitIfOngoing(true).execute().actionGet();
+    }
+
+    @Test
+    public void testEmpty() {
+        ensureGreen();
+        IndicesShardsStoresResponse rsp = client().admin().indices().prepareShardStores().get();
+        assertThat(rsp.getShardStatuses().size(), equalTo(0));
+    }
+
+    @Test
+    public void testBasic() throws Exception {
+        String index = "test";
+        internalCluster().ensureAtLeastNumDataNodes(2);
+        assertAcked(prepareCreate(index).setSettings(Settings.builder()
+                        .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, "2")
+                        .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, "1")
+        ));
+        ensureGreen(index);
+        indexRandomData(index);
+        ensureGreen(index);
+
+        // no unallocated shards
+        IndicesShardsStoresResponse response = client().admin().indices().prepareShardStores(index).get();
+        assertThat(response.getShardStatuses().size(), equalTo(0));
+
+        response = client().admin().indices().shardsStores(Requests.indicesShardsStoresRequest(index).shardStatuses("all")).get();
+        assertThat(response.getShardStatuses().containsKey(index), equalTo(true));
+        ImmutableOpenIntMap<List<IndicesShardsStoresResponse.ShardStoreStatus>> shardStores = response.getShardStatuses().get(index);
+        assertThat(shardStores.values().size(), equalTo(2));
+        for (ObjectCursor<List<IndicesShardsStoresResponse.ShardStoreStatus>> shardStoreStatuses : shardStores.values()) {
+            for (IndicesShardsStoresResponse.ShardStoreStatus shardStoreStatus : shardStoreStatuses.value) {
+                assertThat(shardStoreStatus.getVersion(), greaterThan(-1l));
+                assertThat(shardStoreStatus.getNode(), notNullValue());
+                assertThat(shardStoreStatus.getStoreException(), nullValue());
+            }
+        }
+        ensureGreen(index);
+        logger.info("--> disable allocation");
+        disableAllocation(index);
+        logger.info("--> stop random node");
+        internalCluster().stopRandomNode(new IndexNonMasterNodePredicate(index, internalCluster().getMasterName()));
+        ensureYellow(index);
+        RoutingNodes.UnassignedShards unassigned = clusterService().state().routingNodes().unassigned();
+        response = client().admin().indices().shardsStores(Requests.indicesShardsStoresRequest(index).shardStatuses("yellow")).get();
+        assertThat(response.getShardStatuses().containsKey(index), equalTo(true));
+        ImmutableOpenIntMap<List<IndicesShardsStoresResponse.ShardStoreStatus>> shardStoresStatuses = response.getShardStatuses().get(index);
+        assertThat(shardStoresStatuses.size(), equalTo(unassigned.size()));
```

Comment:
It is technically correct, but only because we have 1 replica and we stopped one node. In general unassigned can have multiple copies of the same shard which may and may not have a store somewhere. Can we just check how many shards have unassigned copies via the routingTable?